File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        #global ans
        ans=[]
        self.count = 0

        def inorder(root, ans):
            if root is None:
                return

            inorder(root.left, ans)
            self.count += 1
            if self.count == k:
                ans.append(root.val)
                return

            inorder(root.right, ans)

            return

        inorder(root, ans)
        return ans[0]

File: Data Structures & Algorithms/path-sum/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    bool  backtrack(TreeNode* root, int &targetSum, int sum){
        if(root==nullptr){
            return false;
        }
        sum+=root->val;
        if(root->left ==nullptr && root->right==nullptr){
            return sum==targetSum;
        }
        bool b1=backtrack(root->left, targetSum, sum);
        bool b2=backtrack(root->right, targetSum, sum);

        
        return b1||b2;

    }


    bool hasPathSum(TreeNode* root, int targetSum) {
        int sum=0;
        

        return backtrack(root, targetSum, sum);
    }
};